### PR TITLE
[Merged by Bors] - Add validation to kdf parameters

### LINKED ISSUE #1906: Insufficient validation of EIP-2335 KDF params
## Description

Lighthouse allows low-security EIP-2335 KDF parameters and we can also accept very high values that might cause a DoS attack.

Please contact me directly for more detail.

### LINKED ISSUE #1907: Unvalidated IV length for EIP-2335
## Description

In our EIP-2335 impl we don't check the IV length. Although the specification doesn't tell us to, it's worth *at least* rejecting a zero-length IV.

Contact me directly for more information.

### PR DESCRIPTION
**Pull request by kirk-baird**
## Issue Addressed

Closes #1906 
Closes #1907 

## Proposed Changes

- Emits warnings when the KDF parameters are two low.
- Returns errors when the KDF parameters are high enough to pose a potential DoS threat.
- Validates AES IV length is 128 bits, errors if empty, warnings otherwise.

## Additional Info

NIST advice used for PBKDF2 ranges https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf. 
Scrypt ranges are based on the maximum value of the `u32` (i.e 4GB of memory)

The minimum range has been set to anything below the default fields.

### CODE REVIEW

File: crypto/eth2_keystore/src/json_keystore/kdf_module.rs [hunk 1]
```diff
@@ -54,6 +55,126 @@ impl Kdf {
...
+                    .checked_mul(params.p)
+                    .ok_or(Error::InvalidScryptParam)?;
+                npr = npr.checked_mul(params.r).ok_or(Error::InvalidScryptParam)?;
+                npr = npr.checked_mul(128).ok_or(Error::InvalidScryptParam)?;
+
+                // Minimum Parameters
+                let default = Scrypt::default_scrypt(vec![0u8; 32]);
+                let default_npr = 128 * default.n * default.p * default.r;
+                if npr < default_npr {
+                    eprintln!("WARN: Scrypt parameters are too weak (n: {}, p: {}, r: {}), we recommend (n: 262,144, p: 1, r: 8)", params.n, params.p, params.r);
```
Comment (paulhauner) [inline crypto/eth2_keystore/src/json_keystore/kdf_module.rs:154]: Nit: we could use the template the `default` values here too.

File: crypto/eth2_keystore/src/json_keystore/kdf_module.rs [hunk 2]
```diff
@@ -54,6 +55,126 @@ impl Kdf {
...
+                if params.c < 262_144 {
+                    if params.c == 0 {
+                        return Err(Error::InvalidPbkdf2Param);
+                    }
+                    eprintln!("WARN: PBKDF2 parameters are too weak, 'c' is {}, we recommend using 262,144", params.c);
+                }
+
+                // Validate `salt` length
+                if params.salt.is_empty() {
+                    return Err(Error::InvalidScryptParam);
```
Comment (paulhauner) [inline crypto/eth2_keystore/src/json_keystore/kdf_module.rs:96]:
```suggestion
                    return Err(Error::InvalidPbkdf2Param);
```

File: crypto/eth2_keystore/src/json_keystore/kdf_module.rs [hunk 3]
```diff
@@ -4,6 +4,7 @@
 //! data structures. Specifically, there should not be any actual crypto logic in this file.
```
Comment (paulhauner) [inline crypto/eth2_keystore/src/json_keystore/kdf_module.rs:4]: I think we're defying this comment by adding the validation here. We could either remove the comment or move these checks into `keystore.rs`. I prefer the latter, personally.

File: crypto/eth2_keystore/src/json_keystore/kdf_module.rs [hunk 4]
```diff
@@ -54,6 +55,126 @@ impl Kdf {
...
+                let default = Scrypt::default_scrypt(vec![0u8; 32]);
+                let default_npr = 128 * default.n * default.p * default.r;
+                if npr < default_npr {
+                    eprintln!("WARN: Scrypt parameters are too weak (n: {}, p: {}, r: {}), we recommend (n: 262,144, p: 1, r: 8)", params.n, params.p, params.r);
+                }
+
+                // Validate `salt` length
+                if params.salt.is_empty() {
+                    return Err(Error::InvalidScryptParam);
+                } else if params.salt.len() < SALT_SIZE / 2 {
```
Comment (paulhauner) [inline crypto/eth2_keystore/src/json_keystore/kdf_module.rs:160]: Nit: we could wrap this check and the next one in a function and use it for pbkdf2 too.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Nice, I like the approach! Only a few small comments.

File: crypto/eth2_keystore/src/json_keystore/kdf_module.rs (hunk 3 above)
Comment (kirk-baird) [reply crypto/eth2_keystore/src/json_keystore/kdf_module.rs:4]: Ooh yep I didn't see that I'll move the functionality to `keystore.rs`

File: crypto/eth2_keystore/src/json_keystore/kdf_module.rs (hunk 4 above)
Comment (kirk-baird) [reply crypto/eth2_keystore/src/json_keystore/kdf_module.rs:160]: Yep can do :)

**Review by paulhauner — COMMENTED** [review]
Perfect!

bors r+

**Review by paulhauner — APPROVED** [review]

**Comment by michaelsproul** [comment]
naughty bors!

bors r-
bors r+

**Pull request closed without merging**